Teach ProxyTorchDispatchMode how to decompose sympy.Expr into known inputs

aorenste opened the pull request:
In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164869
* #164718
* __->__ #164717
* #165266

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @ezyang @EikanWang @jgong5 @wenzhe-nrv

aorenste committed:
Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch
mode is disabled.

[ghstack-poisoned]

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

TODO: Needs better summary
TODO: I need to re-read this before it's ready for review


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste commented on test/distributed/tensor/test_dynamic.py:
```diff
@@ -0,0 +1,63 @@
+class TestDynamic(DTensorTestBase):
+    @with_comms
+    # FIXME: Currently broken for fake tensor cache
+    @parametrize("fake_tensor_cache_enabled", [False])
```
Support for fake_tensor_cache_enabled added in next PR

aorenste commented on torch/distributed/tensor/_sharding_prop.py:
```diff
@@ -166,7 +166,7 @@ def _propagate_tensor_meta_non_cached(
         from torch.fx.experimental.proxy_tensor import disable_proxy_modes_tracing
 
-        with FakeTensorMode(), disable_proxy_modes_tracing():
+        with FakeTensorMode(), disable_proxy_modes_tracing(trace_symbolics=True):
```
All of that and this is the actual fix, lol

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

In sixlib we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

bobrenjc93 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -2461,18 +2492,46 @@ def get_proxy_mode() -> Optional[ProxyTorchDispatchMode]:
+no_sym_dispatch = _NoSymDispatch()
+
+
 def handle_sym_dispatch(
```
Someone (i forgot who) once suggested to me a simple check: run a github search, and if nothing shows up, it’s probably fine:
https://github.com/search?q=handle_sym_dispatch&type=code

bobrenjc93 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -2500,8 +2553,49 @@ def handle_sym_dispatch(
+        #         disable(trace=True):
+        #             ...
+        #
+        new_sym_mode_override = saved_sym_mode_override or mode_unset
```
would we ever expect case like this? 

```
new_proxy_mode:
  disable:
    new_proxy_mode:
      disable:
```

should we add some defensive asserts here for these edge cases we don't expect to support?

bobrenjc93 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -2500,8 +2553,49 @@ def handle_sym_dispatch(
+    """
+    Temporarily disable the current PROXY dispatcher.
+
+    If `trace_symbolics` is True then disables dispatch tracing but continues
```
if you do the above, make sure to rename here as well

bobrenjc93 commented on torch/fx/experimental/sym_node.py:
```diff
@@ -476,19 +476,13 @@ def sym_sum(self, args) -> SymNode:
-                    {},
-                ),
-            )
+        # sym_sum() takes a tuple itself
```
on a secondary note maybe we should type sym_sum in torch/__init__.py to make this more clear for future readers?

aorenste replied:
Just a note since I screwed it up. But I reverted so no point.

bobrenjc93 reviewed (approved):
Accepting to unblock but would like a pass from @bdhirsh before landing if possible

aorenste commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -2479,18 +2504,46 @@ def get_proxy_mode() -> Optional[ProxyTorchDispatchMode]:
+        return no_sym_dispatch
+
+    # Recursively turn the SymNodes in an arg into SymInt/SymFloat/SymBool.
+    def _wrap_node(arg: Any) -> Any:
```
In an earlier version it was necessary (figuring out whether to handle or not was more complicated) but I left it because I like it better. Maybe, given the possibility of a BC break, I should revert it...

bdhirsh commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1549,8 +1567,13 @@ def __torch_dispatch__(
+            torch._C._TorchDispatchModeKey.PROXY
+        )
+        prev_sym_mode = self._global_state.sym_mode_override
+        self.enter_stack.append((prev_proxy_mode, prev_sym_mode))
```
Right now, my understanding is that:

(1) in addition to the mode stack, you have a new global (stashed on the `ProxyTorchDispatchMode` class) that is set whenever we may need to "remember" that we are still doing symint tracing when proxy tracing has been disabled

(2) this allows the code in `disable_proxy_modes_tracing` to unconditionally pop the current proxy mode from the mode stack, since we can use the global independently to determine if we should run `__sym_dispatch__`.

at the risk of nitpicking: what do you think of this vs the alternative of:

(a) have `disable_proxy_modes_tracing(trace_symbolic_shapes=True)` *not* pop the infra mode at all, but set a flag on the `proxy_mode` to mark it as disabled

(b) in the beginning of `ProxyTorchDispatchMode.__torch_dispatch__`, have it no-op if `mode.disabled = True`

I guess I find this easier to reason about (the concept of the global mode stack today and its enter/exit functions are already pretty complicated so worse about making that state more complicated vs the contents of `ProxyMode.__torch_dispatch__`), but let me know if you disagree!

ezyang commented:
Thank you for a clear pull request description

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "Keep tracking SymNodes in the ProxyTorchDispatchMode even while the dispatch"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


mode is disabled.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -406,24 +408,128 @@ def get_proxy_slot(
+        case sympy.Mul():
+            func = operator.mul  # type: ignore[assignment]
+        case _:
+            raise RuntimeError(f"Unhandled sympy type: {type(expr)}")
```
Specifically, don't we have comprehensive sympy to Python type translation already which you can use directly here?

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -406,24 +408,128 @@ def get_proxy_slot(
+      having the inputs available at all - which isn't a guarantee.
+    """
+
+    if (value := tracer.sympy_expr_tracker.get(expr)) is not None:
```
So this is the "allow spurious dependencies" regime right? OK...

aorenste commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -406,24 +408,128 @@ def get_proxy_slot(
+def _build_proxy_for_sym_expr(
+    tracer: _ProxyTracer, expr: sympy.Expr, out: PySymType | None = None
+) -> PySymType | None:
+    """
```
"To handle this we decompose the sympy.Expr and look for the pieces as inputs."

aorenste committed:
Update on "Teach ProxyTorchDispatchMode how to decompose sympy.Expr into known inputs"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

aorenste committed:
Update on "Teach ProxyTorchDispatchMode how to decompose sympy.Expr into known inputs"

In a training library we hit a weird conflict between dtensor, dynamic shapes, and proxy tensor.

The problem is occuring because in sharding_prop we use FakeTensors to compute an operation size (so we don't have to  use the full "real" data). We turn off proxy tracing while we're doing that because we don't want the FakeTensor ops to end up in the graph.  We then use that size when doing later operations.

Normally this is no problem - but when those sizes are dynamic shapes then we have a problem - the proxy tracer wants to track the provenance of all shape operations (`s1*s2`) but since tracing is disabled it doesn't see the operation and when we then use the result shape later on the proxy tracer gets all confused (because the SymNode appeared out of nowhere).

At first we were thinking to never disable shape tracing - but that caused a slew of other downstream problems (lots of code that actually needs the shape tracing to be disabled) so instead we enable having a "sym tracing override" and surgically when we disable proxy tracing we leave shape tracing enabled.

After this change the dtensor embedding is "fixed" but then runs afoul of a FakeTensor cache bug - which is fixed in the next PR.


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]
